gpu/gpu.py

```diff
@@ -0,0 +1,206 @@
+from avocado import Test, skipUnless
+from avocado.utils import build, process, distro, git
+from avocado.utils.software_manager import SoftwareManager
+
+IS_POWER_NV = 'PowerNV' in open('/proc/cpuinfo', 'r').read()
+
+class GPUTests(Test):
+    """
+    GPU test cases make use of sample test code provided by the
+    CUDA toolkil and other generic test suites cover functional tests.
+    Tester to download and setup CUDA stack which will take care of
+    installing nvidia driver too. 
+    """
+    @skipUnless(IS_POWER_NV,
+                "GPU tests are supported only on PowerNV(POWER9)")
+
+    def process_output(self, output):
+       failed_tests = [] 
+       for line in output.decode("utf-8").splitlines():
+           if "fail" in line:
+               failed_tests.append(line)
+       if failed_tests:
+          self.fail("%s" % failed_tests)
+
+    def run_test(self, run_cmd):
+        os.chdir(self.teststmpdir)
+        output = process.system_output(run_cmd, ignore_status=True)
+        self.process_output(output)
+
+    def setUp(self):
+        """
+        Install pre-requisite packages, compile samples and xgemm test code 
+        """
+        smg = SoftwareManager()
+        self.dist = distro.detect()
+        if process.system('lspci| grep NVIDIA', shell=True, 
+                          ignore_status=True):
+            self.cancel("There is NO NVIDIA GPU in the system, tests are\
+                        cancelled.")
+        cuda_rpm = process.system_output('rpm -qa | grep -i cuda', shell=True, 
+                          ignore_status=True).decode()
+        if not cuda_rpm:
+           self.cancel("Please download and setup CUDA toolkit from NVIDIA")
+
+        if self.dist.name not in ['rhel']:
+            self.cancel('Unsupported OS %s' % self.dist.name)
+
+        deps = ['gcc', 'make']
+        for package in deps:
+            if not smg.check_installed(package) and not smg.install(package):
+                self.cancel(
+                    "Fail to install %s required for this test." % (package))
+        cuda_sample  = '/usr/local/cuda/samples/'
+        cuda_exec = 'bin/ppc64le/linux/release/'
+        os.chdir(cuda_sample)
+        build.make(cuda_sample)
+        self.testdir = os.path.join(cuda_sample, cuda_exec)
+
+        for file_name in ['xgemm.c', 'Makefile']:
+            shutil.copyfile(self.get_data(file_name),
+                         os.path.join(self.teststmpdir, file_name))
+        build.make(self.teststmpdir)
+
+    def test_cuda_sample(self):
+        '''
+        GPU cuda sample tests
+        '''
+        self.log.info("test_cuda_sample: Sample cua tests running")
+        os.chdir(self.testdir)
+        for test in os.listdir(self.testdir):
+            self.log.info("Running Sample: %s" % test)
+            test_run = os.path.join(self.testdir, test)
+            process.system_output(test_run, ignore_status=True, sudo=True,\
+                                           shell=True)
```

Comment:
If the output is not required for further processing then `process.system` can be used.

Reply:
Done